extensions/wikia/WallNotifications/WallNotificationControllerBase.class.php

```diff
@@ -0,0 +1,283 @@
+<?php
+
+abstract class WallNotificationControllerBase extends WikiaController {
+	public function __construct() {
+		parent::__construct();
+	}
+
+	public function Index() {
+		global $wgUser, $wgEnableWallExt, $wgEnableForumExt;
+		$loggedIn = $wgUser->isLoggedIn();
+		$suppressWallNotifications = $this->areNotificationsSuppressedByExtensions();
+		if ( $loggedIn && !$suppressWallNotifications ) {
+			$this->addAssets();
+			$this->response->setVal( 'prehide', ( empty( $wgEnableWallExt ) && empty( $wgEnableForumExt ) ) );
+		}
+		$this->response->setVal( 'loggedIn', $loggedIn );
+	}
+
+	public function getDisplayname( $username ) {
+		if ( User::isIP( $username ) ) {
+			return wfMessage( 'oasis-anon-user' )->text();
+		}
+		return $username;
+	}
+
+	public function checkTopic() {
+		// force json format
+		$this->getResponse()->setFormat( 'json' );
+
+		$result = false;
+
+		$topic = $this->getRequest()->getVal( 'query' );
+		if ( !empty( $topic ) ) {
+			/** @var $title title */
+			$title = Title::newFromText( $topic );
+
+			if ( $title instanceof Title ) {
+				$result = (bool)$title->exists();
+			}
+		}
+
+		$this->response->setVal( 'exists', $result );
+	}
+
+	public function updateWiki() {
+		global $wgUser;
+
+		wfProfileIn( __METHOD__ );
+
+		$all = $this->request->getVal( 'notifications' );
+
+		$this->response->setVal( 'user', $wgUser );
+		$this->response->setVal( 'unread', $all[ 'unread' ] );
+		$this->response->setVal( 'read', $all[ 'read' ] );
+
+		wfProfileOut( __METHOD__ );
+	}
+
+	public function Notification() {
+		$wg = F::app()->wg;
+
+		$notify = $this->request->getVal( 'notify' );
+
+		if ( empty( $notify[ 'grouped' ][ 0 ] ) ) {
+			// do not render this notification, it's bugged
+			return;
+		}
+		/** @var WallNotificationEntity $firstEntity */
+		$firstEntity = $notify[ 'grouped' ][ 0 ];
+
+		$data = $firstEntity->data;
+		if ( isset( $data->type ) && ( $data->type === 'ADMIN' || $data->type === 'OWNER' ) ) {
+			$this->forward( __CLASS__, 'NotificationAdmin' );
+			return;
+		}
+
+		$authors = [ ];
+		foreach ( $notify[ 'grouped' ] as $notify_entity ) {
+			$authors[] = [
+				'displayname' => $notify_entity->data->msg_author_displayname,
+				'username' => $notify_entity->data->msg_author_username,
+				'avatar' => AvatarService::renderAvatar(
```

Comment:
!Aware! This part was in GlobalNavigationWallNotificationsController but not in WallNotificationsController